### ISSUE TITLE
Prevent variables from being used if not guaranteed initialization.

### ISSUE THREAD

**Issue by hornta** [eslint/eslint#12575]
**What rule do you want to change?**
https://eslint.org/docs/rules/init-declarations
It may be it's own rule because I couldn't really find a good fit for this.

**Does this change cause the rule to produce more or fewer warnings?**
More

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option

**Please provide some example code that this change will affect:**

Bad: abc may be left uninitialized before used.
```js
function foo() {
  let abc;
  
  if (returnsTrue()) {
    abc = 123; 
  }
  useIt(abc); // Oops, abc is used but may be uninitialized. Possible error?
}

foo();
```

Good: abc is guaranteed to be initialized when used.
```js
function foo() {
  let abc;
  
  if (returnsTrue()) {
    abc = 123; 
  } else {
    return; // this return will make sure abc is initialized before use
  }

  useIt(abc);
}

foo();
```

**What does the rule currently do for this code?**
Nothing.

**What will the rule do after it's changed?**
It will prevent variables from being used if it's not guaranteed that they have been initialized.

**Comment by ilyavolodin** [comment]
I think this rule is getting into a realm of typing (flow/typescript). I think those tools would be better to cover something like this, because ESLint will never be able to check this fully without going through all function calls (which we don't do).

**Comment by hornta** [comment]
I believe it doesn't have to go through all function calls. If it's just sent into a function call or returned from a function it is considered used. I imagined this rule to have to traverse all possible code paths and check for paths that a variable ends up being "used" without having been initialized. As soon as a traversal ends up initializing a variable that path can be closed and it would have to backtrack to the common ancestor... I am not completely familiar with AST traversal but in my head it feels very possible.

**Comment by ilyavolodin** [comment]
Yes, theoretically you can do this. You would need to use [code-path analysis](https://eslint.org/docs/developer-guide/code-path-analysis) which is very complex, but it can check each branch of the code for you. However, here's a perfectly valid JS code that will always return false positives:
```js
// bar.js
function bar() {
  abc = 321;
  return false;
}

//foo.js
const bar = require('./bar');
function foo() {
  var abc;
  
  if (bar()) {
    abc = 123; 
  }
  useIt(abc); // Oops, abc is used but may be uninitialized. Possible error?
}

foo();
```

**Comment by mdjermanovic** [comment]
There is a recently accepted proposal for `no-unassigned-vars` rule in #12497, with the ongoing discussion about its scope, which may include code path analysis. It looks like the suggested enhancements in the discussion there are same as this proposal.
